pkg/lib/notifications.js

```diff
@@ -0,0 +1,161 @@
+ forcing an icon into the navigation menu.
+
+ - title (string, optional)
+
+ A sort, human readable, localized description of the status, suitable
+ for a tooltip.
+
+ - details (JSON value, optional)
+
+ An arbitrary value.  The "System" overview page might monitor a
+ couple of pages for their status and it will use 'details' to display
+ a richer version of the status than possible with just type and
+ title.
+
+Usage:
+
+ import { page_status } from "notifications";
+
+ - page_status.set_own(STATUS)
+
+ Sets the status of the page making the call, completely overwriting
+ the current status.  For example,
+
+    page_status.set_own({
+      type: "info",
+      title: _("Software updates available"),
+      details: {
+        num_updates: 10,
+        num_security_updates: 5
+      }
+    });
+
+    page_status.set_own({
+      type: null
+      title: _("System is up to date"),
+      details: {
+        last_check: 81236457
+      }
+    });
+
+ Calling this function with the same STATUS value multiple times is
+ cheap: only the first call will actually broadcast the new status.
+
+ - page_status.get(PAGE, [HOST])
+
+ Retrieves the current status of page PAGE of HOST.  When HOST is
+ omitted, it defaults to the default host of the calling page.
+
+ PAGE is the same string that Cockpit uses in its URLs to identify a
+ page, such as "system/terminal" or "storage".
+
+ Until the page_status object is fully initialized (see 'valid'
+ below), this function will return 'undefined'.
+
+ - page_status.addEventListener("changed", event => { ... })
+
+ The "changed" event is emitted whenever any page status changes.
+
+ - page_status.valid
+
+ The page_status objects needs to initialize itself asynchronously and
+ 'valid' is false until this is done.  When 'valid' changes to true, a
+ "changed" event is emitted.
+
+*/
+
+import cockpit from "cockpit";
+import deep_equal from "deep-equal";
+
+class PageStatus {
+    constructor() {
+        cockpit.event_target(this);
+        window.addEventListener("storage", event => {
+            if (event.key == "cockpit:page_status") {
+                this.dispatchEvent("changed");
+            }
+        });
+
+        this.cur_own = null;
+
+        this.valid = false;
+        cockpit.transport.wait(() => {
+            this.valid = true;
+            this.dispatchEvent("changed");
+        });
+    }
+
+    get(page, host) {
+        let page_status;
+
+        if (!this.valid)
+            return undefined;
+
+        if (host === undefined)
+            host = cockpit.transport.host;
+
+        try {
+            page_status = JSON.parse(sessionStorage.getItem("cockpit:page_status"));
+        } catch {
+            return null;
+        }
+
+        if (page_status && page_status[host])
+            return page_status[host][page] || null;
+        return null;
+    }
+
+    set_own(status) {
+        if (!deep_equal(status, this.cur_own)) {
```

Comment:
Strawman: Could this always store a JSON encoded value, and then use simple string comparison here? This would avoid a new runtime dependency (not that this hurts much, just an idea).

Reply:
Encoding as JSON is not unique, `{ a: 1, b: 2 }` might be `"{a:1,b:2}"` or `"{b:2,a:1}"`.  There are ways around that, but they also require external code.

Reply:
ack, thanks!